### PR TITLE
[SPARK-26843][MESOS] Use ConfigEntry for hardcoded configs for "mesos" resource manager

### PR DESCRIPTION
**Pull request by HeartSaVioR**
## What changes were proposed in this pull request?

This patch makes hardcoded configs in "mesos" module to use ConfigEntry, avoiding issues on mistake like SPARK-26082.

Please note that there're some changes on type while migrating to ConfigEntry: specifically "comma-separated list on a string" becomes "sequence of strings". While SparkConf smoothly handles on the change (comma-separated list on a string is still supported so backward compatible), there're some methods in utility class (`mesos` package private) to depend on the type change, so this patch also modifies the method signature for them a bit.

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Review by srowen — APPROVED** [review]

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala [hunk 1]
```diff
@@ -34,11 +34,11 @@ import org.apache.spark.internal.Logging
  */
 private[mesos] object MesosSchedulerBackendUtil extends Logging {
   /**
-   * Parse a comma-delimited list of volume specs, each of which
+   * Parse a list of volume specs, each of which
    * takes the form [host-dir:]container-dir[:rw|:ro].
    */
-  def parseVolumesSpec(volumes: String): List[Volume] = {
-    volumes.split(",").map(_.split(":")).flatMap { spec =>
+  def parseVolumesSpec(volumes: Seq[String]): List[Volume] = {
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: @HeartSaVioR .
Why do we change this function signature silently in this `Use ConfigEntry ...` PR?

**Comment by dongjoon-hyun** [comment]
Could you handle
```scala
MesosClusterScheduler.scala:      v => s"$v -Dspark.mesos.driver.frameworkId=${getDriverFrameworkID(desc)}"
MesosClusterScheduler.scala:        logWarning(s"Unknown spark.mesos.appJar.local.resolution.mode $other, using host.")
MesosSchedulerUtils.scala:    sc.conf.remove("spark.mesos.driver.frameworkId")
MesosSchedulerUtils.scala:    System.clearProperty("spark.mesos.driver.frameworkId")
```

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala (hunk 1 above)
Comment (HeartSaVioR) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: This is only used for parsing config value, and we are changing the type of config to `Seq[String]` instead of comma-separated `String`. Same applies for below methods as well.

Similar change was suggested while working on ConfigEntry for others (don't remember which one was...).

I'm not sure changing them would hurt because the object is `mesos` package private.

**Comment by HeartSaVioR** [comment]
Thanks @dongjoon-hyun for finding missing spots! I addressed review comment.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: @HeartSaVioR . You should mention that in the PR description.

Comment (HeartSaVioR) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: TBH I'm not sure we're requiring to mention every incompatible changes for non-public API, but I'm not strongly against it. I'll add it to the PR description.

Please consider that dealing with arbitrary reviewers who have different views on reviewing is not easy. Similar change was done in other PR without any mention and comment. I'd be happy if we have some rules/policies which are across the community in project, and if it doesn't fall into the case, ideally I'd like to see reviewers just kindly and politely suggest/ask it.

Comment (dongjoon-hyun) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: My question was about the **mismatch** between PR title and content itself. If you feel like that, sorry about that. I must be careful. Maybe, `silently` was too strong word?
> Why do we change this function signature silently in this `Use ConfigEntry ...` PR?

BTW, why do you think my comment is about incompatible changes?
> I'm not sure we're requiring to mention every incompatible changes for non-public API

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala [hunk 2]
```diff
@@ -118,6 +143,140 @@ package object config {
...
+    ConfigBuilder("spark.mesos.cluster.retry.wait.max")
+      .intConf
+      .createWithDefault(60) // 1 minute
+
+  private[spark] val ENABLE_FETCHER_CACHE =
+    ConfigBuilder("spark.mesos.fetcherCache.enable")
+      .booleanConf
+      .createWithDefault(false)
+
+  private[spark] val APPLICATION_JAR_LOCAL_RESOLUTION_MODE =
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: nit. Can we use a shorter name like `APP_JAR_LOCAL_RESOLUTION_MODE` like `APP_STATUS_METRICS_ENABLED`?

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala [hunk 3]
```diff
@@ -56,6 +56,26 @@ package object config {
         .createOptional
   }
 
+  private[spark] val CREDENTIAL_PRINCIPAL =
+    ConfigBuilder("spark.mesos.principal")
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Do you think we had better have `.doc` like the previous conf SECRET_FILENAMES at line 50 ~ 56?

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala [hunk 4]
```diff
@@ -118,6 +143,140 @@ package object config {
...
+      .createWithDefault("")
+
+  private[spark] val SLAVE_OFFER_CONSTRAINTS =
+    ConfigBuilder("spark.mesos.constraints")
+      .stringConf
+      .createWithDefault("")
+
+  private[spark] val CONTAINERIZER =
+    ConfigBuilder("spark.mesos.containerizer")
+      .stringConf
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Please add `.checkValues(Set("docker", "mesos"))`.

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala [hunk 5]
```diff
@@ -118,6 +143,140 @@ package object config {
...
+      .createWithDefault("docker")
+
+  private[spark] val ROLE =
+    ConfigBuilder("spark.mesos.role")
+      .stringConf
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Please add `.checkValues(Set("host", "container"))`.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 6]
```diff
@@ -419,22 +419,20 @@ private[spark] class MesosClusterScheduler(
...
     val isLocalJar = desc.jarUrl.startsWith("local://")
-    val isContainerLocal = desc.conf.getOption("spark.mesos.appJar.local.resolution.mode").exists {
+    val isContainerLocal = desc.conf.get(config.APPLICATION_JAR_LOCAL_RESOLUTION_MODE).exists {
       case "container" => true
       case "host" => false
       case other =>
-        logWarning(s"Unknown spark.mesos.appJar.local.resolution.mode $other, using host.")
+        logWarning(s"Unknown ${config.APPLICATION_JAR_LOCAL_RESOLUTION_MODE} $other, using host.")
         false
       }
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: If we have `checkValue`, we can simplify the above logic, too.

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala [hunk 7]
```diff
@@ -118,6 +143,140 @@ package object config {
...
+
+  private[spark] val PROXY_BASE_URL =
+    ConfigBuilder("spark.mesos.proxy.baseURL").stringConf.createOptional
+
+  private[spark] val COARSE_MODE =
+    ConfigBuilder("spark.mesos.coarse").booleanConf.createWithDefault(true)
+
+  private[spark] val COARSE_SHUTDOWN_TIMEOUT =
+    ConfigBuilder("spark.mesos.coarse.shutdownTimeout")
+      .timeConf(TimeUnit.MILLISECONDS)
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Let's add `.checkValue(_ >= 0, s"spark.mesos.coarse.shutdownTimeout must be >= 0")` here.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala [hunk 8]
```diff
@@ -76,15 +76,14 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
...
-  private val maxGpus = conf.getInt("spark.mesos.gpus.max", 0)
+  private val maxGpus = conf.get(MAX_GPUS)
 
-  private val taskLabels = conf.get("spark.mesos.task.labels", "")
+  private val taskLabels = conf.get(TASK_LABELS)
 
   private[this] val shutdownTimeoutMS =
-    conf.getTimeAsMs("spark.mesos.coarse.shutdownTimeout", "10s")
-      .ensuring(_ >= 0, "spark.mesos.coarse.shutdownTimeout must be >= 0")
+    conf.get(COARSE_SHUTDOWN_TIMEOUT).ensuring(_ >= 0, s"$COARSE_SHUTDOWN_TIMEOUT must be >= 0")
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala]: Since we have `checkValue` in config definition, we can remove this `ensuring` here.

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala [hunk 9]
```diff
@@ -118,6 +143,140 @@ package object config {
...
+    ConfigBuilder("spark.mesos.gpus.max")
+      .intConf
+      .createWithDefault(0)
+
+  private[spark] val TASK_LABELS =
+    ConfigBuilder("spark.mesos.task.labels")
+      .stringConf
+      .createWithDefault("")
+
+  private[spark] val SLAVE_OFFER_CONSTRAINTS =
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: I understand this comes from the variable name `slaveOfferConstraints`. But, shall we simply use `CONSTRAINTS` because most Mesos offer configurations are actually about the slave?

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackendSuite.scala [hunk 10]
```diff
@@ -59,8 +59,8 @@ class MesosCoarseGrainedSchedulerBackendSuite extends SparkFunSuite
 
   test("mesos supports killing and limiting executors") {
     setBackend()
-    sparkConf.set("spark.driver.host", "driverHost")
-    sparkConf.set("spark.driver.port", "1234")
+    sparkConf.set(DRIVER_HOST_ADDRESS.key, "driverHost")
+    sparkConf.set(DRIVER_PORT.key, "1234")
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackendSuite.scala]:
```scala
-    sparkConf.set(DRIVER_HOST_ADDRESS.key, "driverHost")
-    sparkConf.set(DRIVER_PORT.key, "1234")
+    sparkConf.set(DRIVER_HOST_ADDRESS, "driverHost")
+    sparkConf.set(DRIVER_PORT, 1234)
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I left a few comments including `.doc`.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala (hunk 1 above)
Comment (HeartSaVioR) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: > My question was about the **mismatch** between PR title and content itself.

I didn't know the reason of the question, and `why` without reason behind makes me feeling that I got non-positive reviews. We would have better understanding of each other when we try to ensure explaining rationalization. 

Btw, the rationalization of method signature change is follow: at the first time I thought I have to add new methods with different signatures (maybe overloaded) but another thought in my mind it would end up letting old methods being not used anymore, and reviewers would ask me to remove them. So flatten the steps into one would be changing signature anyway.

Comment (srowen) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendUtil.scala]: Part of the purpose of these small comments, which can vary across reviewers, is that it's a way for all of us to sync on how to handle the small things.

The change to the method seems fine, and isn't part of any API. I don't think it needs to be part of the title as it is in the description, even. Let's conclude there.

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala (hunk 5 above)
Comment (HeartSaVioR) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Honestly I have no idea about available values. Could you refer where I could find available values? I can't find it in Mesos doc - http://mesos.apache.org/documentation/latest/roles/

Comment (HeartSaVioR) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Ah looks like you're referring `APP_JAR_LOCAL_RESOLUTION_MODE` in below comment. I'll address it.

**Comment by HeartSaVioR** [comment]
Thanks again Dongjoon for detailed and thoughtful reviews! Addressed review comments.

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala [hunk 11]
```diff
@@ -118,6 +148,143 @@ package object config {
...
+  private[spark] val RETAINED_DRIVERS =
+    ConfigBuilder("spark.mesos.retainedDrivers").intConf.createWithDefault(200)
+
+  private[spark] val CLUSTER_RETRY_WAIT_MAX_SECONDS =
+    ConfigBuilder("spark.mesos.cluster.retry.wait.max")
+      .intConf
+      .createWithDefault(60) // 1 minute
+
+  private[spark] val ENABLE_FETCHER_CACHE =
+    ConfigBuilder("spark.mesos.fetcherCache.enable")
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: While you are here, can we copy a documentation description in `running-on-mesos.md` to `.doc` for the new configurations? At least, for the already documented ones, I think we can have.

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Yes. It seems that I clicked a wrong one.

**Comment by dongjoon-hyun** [comment]
@clems4ever This PR is almost done. I'm waiting for a few minor `.doc` additions to sync with documentation.

**Comment by clems4ever** [comment]
Alright @dongjoon-hyun, I'll keep an eye on this PR. Thanks for letting me know.

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackendSuite.scala [hunk 12]
```diff
@@ -80,9 +81,9 @@ class MesosFineGrainedSchedulerBackendSuite
   }
 
   test("Use configured mesosExecutor.cores for ExecutorInfo") {
-    val mesosExecutorCores = 3
+    val mesosExecutorCores = 3.0
```
Comment (felixcheung) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackendSuite.scala]: why is this change needed?

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 13]
```diff
@@ -419,22 +419,17 @@ private[spark] class MesosClusterScheduler(
...
 
   private def getDriverUris(desc: MesosDriverDescription): List[CommandInfo.URI] = {
-    val confUris = List(conf.getOption("spark.mesos.uris"),
-      desc.conf.getOption("spark.mesos.uris"),
-      Some(desc.conf.get(SUBMIT_PYTHON_FILES).mkString(","))).flatMap(
-      _.map(_.split(",").map(_.trim))
-    ).flatten
+    val confUris = (conf.get(config.URIS_TO_DOWNLOAD) ++
+      desc.conf.get(config.URIS_TO_DOWNLOAD) ++
+      desc.conf.get(SUBMIT_PYTHON_FILES)).toList
```
Comment (felixcheung) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: I'm not sure this is exactly the same? `_.split(",").map(_.trim)` - before `URIS_TO_DOWNLOAD` should be a comma separated list (probably not intentional, but it works) and `URIS_TO_DOWNLOAD` & `SUBMIT_PYTHON_FILES` could have whitespaces

**Review by felixcheung — COMMENTED** [review]
LG mostly, 2 questions

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 13 above)
Comment (HeartSaVioR) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Element values will be trimmed as same as it was - ConfigBuilder handles it:

https://github.com/apache/spark/blob/dc46fb77ba70709374b4e5feec60e59e33a1ea52/core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala#L123-L126

https://github.com/apache/spark/blob/dc46fb77ba70709374b4e5feec60e59e33a1ea52/core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala#L48-L50

https://github.com/apache/spark/blob/dc46fb77ba70709374b4e5feec60e59e33a1ea52/core/src/main/scala/org/apache/spark/util/Utils.scala#L2619-L2621

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackendSuite.scala (hunk 12 above)
Comment (HeartSaVioR) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackendSuite.scala]: The type of EXECUTOR_CORES is actually `double` instead of `int` or `long` and if my memory is right compiler complained. If we don't want to touch this we may need to change below line to `conf.set(EXECUTOR_CORES.key, mesosExecutorCores.toString)`

Comment (srowen) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackendSuite.scala]: The type of a number of cores is logically an integer and can be used where a double is expected. In a few cases a cast may be needed but doesn't require thinking of number of cores as a float

Comment (HeartSaVioR) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackendSuite.scala]: I'm honestly not a expert on this but the documentation of `spark.mesos.mesosExecutor.cores` clarify that it can receive floating point number so the type still makes sense.

https://github.com/apache/spark/blob/master/docs/running-on-mesos.md

The default value is also explained as `1.0` instead of 1. I'll update the config entry to reflect the sync.

**Comment by HeartSaVioR** [comment]
Documentation addressed. Please note that I only copied the contents from `Spark Properties` table. For others only referred or mentioned I didn't touch them to avoid bad explanation.

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackendSuite.scala (hunk 12 above)
Comment (srowen) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackendSuite.scala]: Oh right mesos can specify fractional cores. OK this makes sense then.

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala [hunk 14]
```diff
@@ -118,6 +150,221 @@ package object config {
...
+        "If your label includes a colon or comma, you can escape it with a backslash. " +
+        "Ex. key:value,key2:a\\:b.")
+      .stringConf
+      .createWithDefault("")
+
+  private[spark] val CONSTRAINTS =
+    ConfigBuilder("spark.mesos.constraints")
+      .doc("Attribute-based constraints on mesos resource offers. By default, all resource " +
+        "offers will be accepted. This setting applies only to executors. Refer to Mesos " +
+        "Attributes & Resources doc for more information on attributes. " +
```
Comment (dongjoon-hyun) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Hi, @HeartSaVioR .
Could you remove the followings line 338 ~ 347? Since it says `Refer to Mesos ...` already, it looks okay to remove the rest.

Comment (HeartSaVioR) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/config.scala]: Totally agreed. Just addressed it.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you, @HeartSaVioR , @srowen , @felixcheung !

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!
